### PR TITLE
gh-102780: Fix uncancel() call in asyncio timeouts

### PR DESCRIPTION
**Pull request by kristjanvalur**
Record the previous cancellation state when entering `asyncio.timeout.Timeout`
This allows us to use `asyncio.timeout` even when handling a `CancelledError`
exception.   The `Timeout` context manager will still not handle a CancelledError
if a _new_ cancellation request is delivered in its scope.

- Fix Timeout context manager
- Add regression tests
- Add documentiation, clarifying when it is appropriate to call `Task.uncancel()`

This needs to be backported to 3.11

* Issue: gh-102780

### CODE REVIEW

**Comment by gvanrossum** [comment]
Thanks! This looks like the crux of the matter, and beautifully executed. I'll properly review it later this weekend.

File: Doc/library/asyncio-task.rst
Comment (gvanrossum) [inline Doc/library/asyncio-task.rst:307]: Could I ask you to also rephrase "Most code can safely ignore :exc:`asyncio.CancelledError`" above? This vaguely seems like it would allow
```py
except CancelledError:
    pass  # Ignore CancelledError
```
which is the opposite of what it's trying to say.

Maybe say that most code doesn't need to be aware of it?

File: Doc/library/asyncio-task.rst
Comment (gvanrossum) [inline Doc/library/asyncio-task.rst]: Drop "such"
```suggestion
However, in cases when suppressing :exc:`asyncio.CancelledError` is
```

File: Lib/asyncio/timeouts.py
Comment (gvanrossum) [inline Lib/asyncio/timeouts.py:111]: This is a nice improvement that maybe ought to be called out in the commit message? (If you agree I can handle that when I merge it.)

**Review by gvanrossum — APPROVED** [review]
I only really have doc nits, plus a feeling that maybe possibly TaskGroup could have the same problem?

I'm asking @kumaraditya303 to also review this, to make sure I didn't miss anything.

File: Lib/asyncio/timeouts.py
Comment (kumaraditya303) [reply Lib/asyncio/timeouts.py:111]: Isn't this done by default? I see the same traceback with and without this on this example:

```py
import asyncio

async def func():
    await asyncio.sleep(1)

async def main():
    async with asyncio.timeout(0.1):
        await func()

asyncio.run(main())
```

File: Doc/library/asyncio-task.rst
Comment (puntonim) [reply Doc/library/asyncio-task.rst]: Mind the typo in `supprssing`

File: Lib/asyncio/timeouts.py
Comment (gvanrossum) [reply Lib/asyncio/timeouts.py:111]: Not quite. When using `raise exc from ...` the `CancelledError` instance gets stored in `exc.__cause__`, while without it, it gets stored in `exc.__context__`, and the connecting phrase in the output is different:

`__cause__`:
```
The above exception was the direct cause of the following exception:
```

`__context__`:
```
During handling of the above exception, another exception occurred:
```

(If both are set it follows `__cause__`.

Comment (kristjanvalur) [reply Lib/asyncio/timeouts.py:111]: Precicely.  I has bothered me for a while that timeouts happen with the latter message, as if somehow, there was an error during exception handling.  I wanted to use the opportunity to roll that fix in :)

File: Doc/library/asyncio-task.rst
Comment (kristjanvalur) [reply Doc/library/asyncio-task.rst:307]: How about the above?

File: Lib/asyncio/timeouts.py
Comment (kristjanvalur) [reply Lib/asyncio/timeouts.py:111]: > This is a nice improvement that maybe ought to be called out in the commit message? (If you agree I can handle that when I merge it.)

Thank you, agreed.

Comment (kumaraditya303) [reply Lib/asyncio/timeouts.py:111]: > Not quite. When using raise exc from ... the CancelledError instance gets stored in exc.__cause__, while without it, it gets stored in exc.__context__, and the connecting phrase in the output is different:

Ah, right! I missed that. Adding a test would be nice.

**Review by kumaraditya303 — APPROVED** [review]
I would like to have a test for the exception chaining else LGTM. I leave the docs for Guido to review. 

@gvanrossum Feel free to merge as you like.

File: Doc/library/asyncio-task.rst
Comment (AlexWaygood) [inline Doc/library/asyncio-task.rst]: Strictly speaking _all_ exceptions are "a subclass" of `BaseException` — this wording might make it clearer what's going on here:

```suggestion
clean-up is complete. :exc:`asyncio.CancelledError` directly subclasses
```

File: Doc/library/asyncio-task.rst
Comment (AlexWaygood) [inline Doc/library/asyncio-task.rst]: This link doesn't work. I don't think we need it to work, since we link to `uncancel()` immediately above, but it will lead to a Sphinx warning being generated if the docs are built in "nitpicky mode" (#101100). Better to do this instead:

```suggestion
truly desired, it is necessary to also call ``uncancel()`` to completely
```

Comment (kristjanvalur) [reply Doc/library/asyncio-task.rst]: Thanks, its been a long time since I worked with Python docs.

File: Lib/asyncio/timeouts.py
Comment (kristjanvalur) [reply Lib/asyncio/timeouts.py:111]: I'll add that in a few hours.

**Review by AlexWaygood — COMMENTED** [review]
The docs read okay to me now, but I'll let Guido or Kumar merge since I'm not an asyncio expert!

**Review by gvanrossum — COMMENTED** [review]
Needs a news entry, otherwise LGTM.

File: Misc/NEWS.d/next/Library/2023-03-22-16-15-18.bpo-102780.NEcljy.rst
Comment (AlexWaygood) [inline Misc/NEWS.d/next/Library/2023-03-22-16-15-18.bpo-102780.NEcljy.rst]:
```suggestion
:func:`asyncio.timeout` now works reliably even when performing cleanup due
to task cancellation.  Previously the context manager could raise a
:exc:`asyncio.CancelledError` instead of an :exc:`asyncio.TimeoutError` in such cases.
```

File: Misc/NEWS.d/next/Library/2023-03-22-16-15-18.gh-issue-102780.NEcljy.rst
Comment (AlexWaygood) [inline Misc/NEWS.d/next/Library/2023-03-22-16-15-18.gh-issue-102780.NEcljy.rst]:
```suggestion
:func:`asyncio.timeout` now works reliably even when performing cleanup due
to task cancellation.  Previously the context manager could raise a
:exc:`asyncio.CancelledError` instead of an :exc:`asyncio.TimeoutError` in such cases.
```

Comment (kristjanvalur) [reply Misc/NEWS.d/next/Library/2023-03-22-16-15-18.gh-issue-102780.NEcljy.rst]: Thanks, I already did something similar.  Been a while since I grappled with ReST and Python.

**Review by gvanrossum — APPROVED** [review]

**Comment by graingert** [comment]
does this need a corresponding change in asyncio.TaskGroup? eg here https://github.com/python/cpython/blob/e90036c9bdf25621c15207a9cabd119fb5366c27/Lib/asyncio/taskgroups.py#L82

**Comment by kristjanvalur** [comment]
Interesting.  It would be an issue if we entered a TaskGroup while cancellation state was not 0, e.g. during a finally of an outer cancel.   I'm not qualified, though, to judge if that would ever be a real issue.

**Pull request merged**
